WalletWasabi.Tests/UnitTests/BlockchainAnalysis/PubKeyReuseAnonScoreTests.cs

```diff
@@ -0,0 +1,154 @@
+			var analyser = ServiceFactory.CreateBlockchainAnalyzer();
+			var km = ServiceFactory.CreateKeyManager();
+			var key = BitcoinFactory.CreateHdPubKey(km);
+			var tx = BitcoinFactory.CreateSmartTransaction(
+				0,
+				Enumerable.Repeat(Money.Coins(1m), 9),
+				new[] { (Money.Coins(1.1m), 100, key), (Money.Coins(1.2m), 100, key), (Money.Coins(1.3m), 100, key), (Money.Coins(1.4m), 100, key) },
+				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet, BitcoinFactory.CreateHdPubKey(km)) });
+
+			analyser.Analyze(tx);
+
+			Assert.All(tx.WalletInputs, x => Assert.Equal(1, x.HdPubKey.AnonymitySet));
+			Assert.Equal(1, tx.WalletOutputs.First().HdPubKey.AnonymitySet);
+		}
+
+		[Fact]
+		public void InputSideAddressReuseHaveNoConsolidationPunishmentInSelfSpend()
+		{
+			var analyser = ServiceFactory.CreateBlockchainAnalyzer();
+			var km = ServiceFactory.CreateKeyManager();
+			var key = BitcoinFactory.CreateHdPubKey(km);
+			var tx = BitcoinFactory.CreateSmartTransaction(
+				0,
+				Enumerable.Empty<Money>(),
+				new[] { (Money.Coins(1.1m), 100, key), (Money.Coins(1.2m), 100, key), (Money.Coins(1.3m), 100, key), (Money.Coins(1.4m), 100, key) },
+				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet, BitcoinFactory.CreateHdPubKey(km)) });
+
+			analyser.Analyze(tx);
+
+			Assert.All(tx.WalletInputs, x => Assert.Equal(100, x.HdPubKey.AnonymitySet));
+			Assert.Equal(100, tx.WalletOutputs.First().HdPubKey.AnonymitySet);
+		}
+
+		[Fact]
+		public void InputSideAddressReuseHaveNoConsolidationPunishmentInCoinJoin()
+		{
+			var analyser = ServiceFactory.CreateBlockchainAnalyzer();
+			var km = ServiceFactory.CreateKeyManager();
+			var key = BitcoinFactory.CreateHdPubKey(km);
+			var tx = BitcoinFactory.CreateSmartTransaction(
+				9,
+				Enumerable.Repeat(Money.Coins(1m), 9),
+				new[] { (Money.Coins(1.1m), 100, key), (Money.Coins(1.2m), 100, key), (Money.Coins(1.3m), 100, key), (Money.Coins(1.4m), 100, key) },
+				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet, BitcoinFactory.CreateHdPubKey(km)) });
+
+			analyser.Analyze(tx);
+
+			Assert.All(tx.WalletInputs, x => Assert.Equal(100, x.HdPubKey.AnonymitySet));
+			Assert.Equal(109, tx.WalletOutputs.First().HdPubKey.AnonymitySet);
+		}
+
+		[Fact]
+		public void InputOutputSideAddress()
+		{
+			// If there's reuse in input and output side, then output side didn't gain, nor lose anonymity.
+			var analyser = ServiceFactory.CreateBlockchainAnalyzer();
+			var key = BitcoinFactory.CreateHdPubKey(ServiceFactory.CreateKeyManager());
+			var tx = BitcoinFactory.CreateSmartTransaction(
+				9,
+				Enumerable.Repeat(Money.Coins(1m), 9),
+				new[] { (Money.Coins(1.1m), 100, key) },
+				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet, key) });
+
+			analyser.Analyze(tx);
+
+			Assert.All(tx.WalletInputs, x => Assert.Equal(100, x.HdPubKey.AnonymitySet));
+			Assert.Equal(100, tx.WalletOutputs.First().HdPubKey.AnonymitySet);
+		}
+
```

Comment:
i think we also need another test for when there's reuse in other outputs (e.g. past transactions), since that would affect the calculation.

if i'm not mistaken this could should be failing, but i think the floor should be 1, see rationale in `AnalyzeCoinJoin` comments below.

```suggestion
		[Fact]
		public void InputOutputSideAddress()
		{
			// If there's reuse in input and output side, then output side didn't gain, nor lose anonymity.
			var analyser = ServiceFactory.CreateBlockchainAnalyzer();
			var key = BitcoinFactory.CreateHdPubKey(ServiceFactory.CreateKeyManager());
			var tx = BitcoinFactory.CreateSmartTransaction(
				9,
				Enumerable.Repeat(Money.Coins(1m), 9),
				new[] { (Money.Coins(1.1m), 100, key) },
				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet, key) });

			analyser.Analyze(tx);

			Assert.All(tx.WalletInputs, x => Assert.Equal(100, x.HdPubKey.AnonymitySet));
			Assert.Equal(100, tx.WalletOutputs.First().HdPubKey.AnonymitySet);
		}

		[Fact]
		public void InputOutputSidePreviouslyUsedAddress()
		{
			// If there's reuse in input and output side, then output side didn't gain, nor lose anonymity.
			var analyser = ServiceFactory.CreateBlockchainAnalyzer();
			var reuse = BitcoinFactory.CreateHdPubKey(ServiceFactory.CreateKeyManager());
			var tx = BitcoinFactory.CreateSmartTransaction(
				9,
				Enumerable.Repeat(Money.Coins(1m), 9),
				new[] { (Money.Coins(1.1m), 100, key) },
				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet, reuse) });

			reuse.AnonymitySet = 30;

			analyser.Analyze(tx);
			
			Assert.All(tx.WalletInputs, x => Assert.True(x.HdPubKey.AnonymitySet < 30));
			Assert.True(tx.WalletOutputs.First().HdPubKey.AnonymitySet < 30);
			// also this? // Assert.True(reuse.AnonymitySet < 30);
		}

```

Reply:
Implemented: [`9f99b17` (#4724)](https://github.com/zkSNACKs/WalletWasabi/pull/4724/commits/9f99b1791b56a9cd0ce381207f48508403474888)